Database provider

SanaKhan1997 opened the pull request:
Closes #70.

holtzmak committed:
Minor changes to mutability

holtzmak committed:
Add ATR preview widget card, ViewModel, and hookup

holtzmak committed:
Fixed test structure to match lib

holtzmak committed:
Add ATR test tools

holtzmak committed:
Add ATR preview test

holtzmak committed:
Adjusted ATR Preview formatting

holtzmak committed:
Run flutter pub upgrade

SanaKhan1997 committed:
Added local and remote database

SanaKhan1997 committed:
added local and remote database provider

holtzmak committed:
Refactor constructors

holtzmak committed:
Refactor test

holtzmak committed:
Add DatabaseService placeholder and TODOs

SanaKhan1997 committed:
added map to transporter_info

SanaKhan1997 committed:
added map to transporter_info

holtzmak commented on app/lib/core/models/transporter_info.dart:
```diff
@@ -47,4 +47,23 @@ class TransporterInfo {
 
   List<String> driverNames() => List.unmodifiable(_driverNames);
+
+  Map<String, dynamic> toMap() {
```
Looks like we'll be reusing this in almost every ATR-related object.

Could we use a library like [this one](https://api.dart.dev/stable/2.8.4/dart-convert/dart-convert-library.html) to make all the `toMap`? 

What worries me about all these functions is that we could miss a field in them, like if you missed `driverNames` below, there's no way you would know immediately until a test or something failed. I read a little on dart forums and it looks like there's no automatic `toMap` for objects yet.

holtzmak commented on app/lib/core/services/database_provider/db_services.dart:
```diff
@@ -0,0 +1,212 @@
+  var networkStatus;
+  MySqlConnection dbConnect;
+
+  Future<bool> checkConnectivity() async {
```
I really hope there's either a plugin for this, or Firestore provides this.

SanaKhan1997 replied:
I used connectivity plugin for this so it should work with firestore as well

holtzmak commented on app/lib/core/services/database_provider/db_services.dart:
```diff
@@ -0,0 +1,212 @@
+  MySqlConnection dbConnect;
+
+  Future<bool> checkConnectivity() async {
+    var connectivityResult = await (Connectivity().checkConnectivity());
```
If using `await`, be sure to wrap it in a `try-catch` or some other Future-istic code, especially in the database (we don't do this with tests because they're supposed to work)

holtzmak commented on app/lib/core/services/database_provider/db_services.dart:
```diff
@@ -0,0 +1,212 @@
+          newRecord.contingencyInfo.potentialHazards()
+        ]);
+
+    await dbConnect.query('''
```
Each of these should be their own function, it will help us to narrow down the point of failure in the future if something breaks (like we change an object somewhere, or add a field)

holtzmak commented on app/pubspec.lock:
```diff
@@ -648,5 +711,5 @@ packages:
 sdks:
-  dart: ">=2.10.0 <2.11.0"
-  flutter: ">=1.16.0 <2.0.0"
+  dart: ">=2.10.2 <2.11.0"
```
- [x] Delete `pubspec.lock` and re-run `flutter pub upgrade `to regenerate it.

holtzmak replied:
Actually, one of the new additions to pubspec.yaml requires a minimum dart and flutter version. This is fine

holtzmak commented on app/pubspec.yaml:
```diff
@@ -34,6 +34,9 @@ dependencies:
   path_provider: ^1.5.1
   pdf: ^2.0.0
   provider: ^4.3.2+3
+  mysql1: ^0.17.1
```
```suggestion
  # TODO: #95 or #130. Remove once we've decided on a database (rerun flutter pub get)
  mysql1: ^0.17.1
```

holtzmak reviewed (changes requested):
This is a very good first shot at the database! Sorry we keep changing our minds about what database to use (it's good we're doing it now than later at least).

What do you think about making an abstract class for all databases? Like:
```
abstract class DatabaseInterface{
   Future<Record> getRecord(Record rec);
   // and other functions for every query we could make
}
```
That `DatabaseService` or whatever would implement? I think we should extract all your functions and make this now for future developers (or even us if we make changes again).

- [x] Add abstract DatabaseService class that all database services implement


For all my other comments, since this code is experimental and using MySQL, it's likely what you've written will be refactored a lot. Up to you how many changes you want to make.

MummenRider commented on app/lib/core/services/database_provider/local_db_services.dart:
```diff
@@ -0,0 +1,63 @@
+import 'package:path_provider/path_provider.dart';
+
+class LocalDB {
+  static Database _database;
+  static const String TRANSPORTER = 'TransporterInfo';
+  static const String SHIPPER = 'ShipperInfo';
+  static const String DELIVERY = 'DeliveryInfo';
+  static const String FWR = 'FwrInfo';
+  static const String VEHICLE = 'VehicleLoadingInfo';
+  static const String CONTINGENCY = 'ContingencyPlan';
+  static const String ACKNOWLEDGMENTS = 'AckInfo';
+  static const String DB_NAME = 'humane_transport.db';
```
Why do we nee to define the string here? Can't we create a model for this instead?

holtzmak replied:
Oh yeah, these constants can be in their respective classes

MummenRider reviewed (commented):
Queries are pretty similar to Firestore, which is good! It looks like you created a bunch of queries, but did you try it to see if it works?

holtzmak committed:
add test_lib dir

holtzmak committed:
Added history test for retieve one ATR

holtzmak committed:
Add history test to actions

holtzmak committed:
Fix main_test

Moved test_lib because test package didn't like it in global directory. Added beginnings of test service locator.

holtzmak committed:
Add toString methods to ATR

SanaKhan1997 committed:
Update README.md (#147)

holtzmak committed:
Moved utility widgets to ui/widgets/utility

holtzmak committed:
Create single history screen

holtzmak committed:
Add back button to single history screen

holtzmak committed:
Add single history screen test

holtzmak committed:
Add rest of ATR toString methods

These are not a permanent solution, and will likely move to UI code

holtzmak committed:
Add rest of tests to ATR display test

holtzmak commented:
@SanaKhan1997 as discussed in backlog refinement, a lot of this code needs to be replaced with Firestore equivalent calls. So, feel free to ignore comments that ask you to refactor code, because this work will be replaced.

Please still do make the abstract class and fix the `pubspec` file.

MummenRider committed:
Dialog service  (#157)

* Dialog service is added in this commit

* Used optional instead of null values

* removed alert dialog widget

MummenRider committed:
Response class (#159)

* Added Response class

* Updated Response type as per requested

holtzmak committed:
Run flutter pub upgrade

holtzmak committed:
Big refactoring of file names and locations

Did this to make things consistent. Also gearing up for design 2 implementation.

holtzmak committed:
Refactor tests with code changes

holtzmak committed:
Pre-commit run all files

holtzmak committed:
Fix HistoryScreen test after refactoring

holtzmak committed:
Changes to how navigation is done

holtzmak committed:
Updated integration tests names

holtzmak committed:
Add test screen for navigation without bottom nav

holtzmak committed:
Change Active to ActiveScreen

holtzmak committed:
Add route generator comments for nav bar persistence

holtzmak committed:
Add navigateTo, navigateBackUntil functions

holtzmak committed:
Add NavService test

Shows how to use functions. Shows how to pass arguments.

holtzmak committed:
Changed NavServiceTest expectations to be explicit

holtzmak committed:
Upgrade ATRPreviews to use NavService

holtzmak committed:
Rename package new_travel to new

holtzmak committed:
Address inconsistencies with NavService and NavDrawer

holtzmak committed:
Update HistoryScreen test to look for particular ATRPreview

holtzmak committed:
Specified that NavigationService is just for main navigator

holtzmak committed:
Add testLocator to ATRDisplay test

holtzmak committed:
Commit to using NavService everywhere for main navigator routing

Except ATRDisplay, to be done in another commit

SanaKhan1997 committed:
Added abstract database

SanaKhan1997 committed:
Added abstract database

SanaKhan1997 committed:
flutter upgrade

SanaKhan1997 committed:
flutter upgrade

SanaKhan1997 committed:
temp

SanaKhan1997 committed:
temp

holtzmak commented:
Because @SanaKhan1997 rebase got messed up. I opened a [new branch](https://github.com/holtzmak/Humane-Transport/tree/add-abstract-database-interface) with her picked changes. This will either be opened in a new PR, or I will overwrite this branch.

 Sana, let me know that the branch has all your necessary changes please.
